solarkraft/src/verify_js.ts

```diff
@@ -0,0 +1,329 @@
+                    .instance,
+            persistent: () =>
+                this.tx.oldStorage.get(contractId, emptyContractStorage())
+                    .persistent,
+            temporary: () =>
+                this.tx.oldStorage.get(contractId, emptyContractStorage())
+                    .temporary,
+        }
+    }
+}
+
+/**
+ * The environment available to monitor functions, when the transaction succeeds.
+ *
+ * `storage()` returns the storage of the contract at the end of the transaction.
+ * `oldStorage()` returns the storage of the contract at the beginning of the transaction.
+ */
+export class Env extends AbstractEnv {
+    storage(contractId: string = this.tx.contractId): EnvStorage {
+        return {
+            instance: () =>
+                this.tx.storage.get(contractId, emptyContractStorage())
+                    .instance,
+            persistent: () =>
+                this.tx.storage.get(contractId, emptyContractStorage())
+                    .persistent,
+            temporary: () =>
+                this.tx.storage.get(contractId, emptyContractStorage())
+                    .temporary,
+        }
+    }
+    oldStorage(contractId: string = this.tx.contractId): EnvStorage {
+        return {
+            instance: () =>
+                this.tx.oldStorage.get(contractId, emptyContractStorage())
+                    .instance,
+            persistent: () =>
+                this.tx.oldStorage.get(contractId, emptyContractStorage())
+                    .persistent,
+            temporary: () =>
+                this.tx.oldStorage.get(contractId, emptyContractStorage())
+                    .temporary,
+        }
+    }
+}
+
+/* A simple boolean expression language */
+
+type ConditionSome = { type: 'some'; conditions: Condition[] }
+type ConditionEvery = { type: 'every'; conditions: Condition[] }
+type Condition = ConditionSome | ConditionEvery | boolean
+
+/** Logical OR of the given conditions. */
+export function some(...args: Condition[]): ConditionSome {
+    return { type: 'some', conditions: args }
+}
+
+/** Logical AND of the given conditions.  */
+export function every(...args: Condition[]): ConditionEvery {
+    return { type: 'every', conditions: args }
+}
+
+/** Evaluate a condition to a boolean. */
+export function evaluateCondition(condition: Condition): boolean {
+    if (typeof condition === 'boolean') {
+        return condition
+    }
+    switch (condition.type) {
+        case 'every':
+            return condition.conditions.every((x) => evaluateCondition(x))
+        case 'some':
+            return condition.conditions.some((x) => evaluateCondition(x))
+    }
+}
+
+/* Macros / monitor helper functions */
+
+/**
+ * Return a condition that checks if the given token amount has been transferred from one account to another.
+ */
+export function tokenTransferred(
+    env: Env,
+    token: string,
+    from: string,
+    to: string,
+    amount: number
```

Comment:
would not it be better to use big integers right from the start? Otherwise, we would not be able to test amount that do not fit into `number`.

Reply:
True, I also encountered this while working on an additional case study.

I have to do a bit of digging to find out where we introduce `number`. `js-stellar-sdk` actually should give us `bigint`s, so they are probably lost somewhere in serialization.

Reply:
Since I already have another PR wedged inbetween, I'm fixing this in a separate one: #140